src/anm/analytics.js

```diff
@@ -0,0 +1,79 @@
+var engine = require('engine'),
+    utils = require('./utils.js');
+
+var Analytics = function () {
+    var self = this,
+        supportSendBeacon = !!navigator.sendBeacon,
+        timeout = 1000,
+        beacon = null,
+        animatronUrl = utils.makeApiUrl('/analytics/player');
+
+    self.enabled = animatronUrl != null && animatronUrl.indexOf('animatron-test') >= 0;
+    self.queue = [];
+
+    var event = function () {
+        if (self.queue.length > 0) {
+            var data = JSON.stringify(self.queue);
+            self.queue = [];
+
+            if (supportSendBeacon) {
+                navigator.sendBeacon(animatronUrl, data);
+            } else {
+                sendViaGif(data);
+            }
+        }
+        if (!supportSendBeacon) {
+            setTimeout(event, timeout);
+        }
+    };
+
+    var sendViaGif = function (data) {
+        if (!beacon) {
+            beacon = engine.createStatImg();
+        }
+        beacon.src = animatronUrl + '?data=' + encodeURIComponent(data);
+    };
+    if (self.enabled) {
+        if (!supportSendBeacon) {
+            engine.onDocReady(event);
```

Comment:
Vova, could it be that you misused `setTimeout` instead of `setInterval`? If you'd use `setInterval` instead, things'd start to get a little more sense. 

The problem is not in checking if array is empty once more. And it's not an event loop. Once again.
An example with browser without `sendBeacon`.

1. Browser or tab opened by user with this page. 
1. Browser loads scripts.
1. `Analytics` object created. Queue is empty. Postpone `event` for one second. (Imagine one second here).
1. Somewhere here browser starts to build DOM.
1. One or two second pass.
1. `event` function is called thanks to timeout. Queue is empty.
1. DOM finished construction.
1. ...
1. Some time passes, Players created.
1. ...
1. A lot time passes, during that time user does actions, fills `queue` with events, now it's not empty.
1. ...
1. User decides to close tab or a browser or "crash" happens.
1. Probably (you say, not guaranteed) `window.unload` handler is called. If it did, all the collected events flush to the server.
1. Fin.

How the call at time `6.` affects _safety_? If `window.unload` call is not guaranteed, in case of failure, nothing will be sent anyway. Timeouts complicate code.